File: jetbrains/src/main/kotlin/deplens/utils/inlay/NpmInlayUtils.kt
package deplens.utils.inlay

import com.intellij.codeInsight.hints.declarative.InlayTreeSink
import com.intellij.openapi.diagnostic.logger
import com.intellij.psi.PsiFile
import deplens.common.I18nKey
import deplens.common.Result
import deplens.utils.service.GithubRepoInfoService
import deplens.utils.I18n
import deplens.utils.ProgressUtils
import deplens.utils.service.NpmPkgInfoService
import deplens.utils.UiUtils

object NpmInlayUtils {

    private val LOG = logger<NpmInlayUtils>()

    fun addNpmDepInlay(file: PsiFile, sink: InlayTreeSink, pkg: String, offset: Int) {
        val project = file.project
        val vFile = file.virtualFile
        val npmRes = NpmPkgInfoService.getPackageInfo(pkg)

        when (npmRes.result) {
            Result.NONE -> {
                if (NpmPkgInfoService.hasFailure(pkg) && !NpmPkgInfoService.isRequestRunning(pkg)) {
                    UiUtils.addInlay(
                        sink,
                        offset,
                        I18n.message(I18nKey.failedNpmMeta),
                        retryToken = "npm:$pkg"
                    )
                    return
                }
                ProgressUtils.runBackground(file.project, "DepLens: Fetch npm $pkg") {
                    try {
                        NpmPkgInfoService.fetchPackageInfo(pkg) {
                            var dispatchedGithubFetch = false

                            // 请求完npm元数据
                            val npmRes = NpmPkgInfoService.getPackageInfo(pkg)

                            if (npmRes.result == Result.SUCCESS) {
                                val url = (npmRes.data?.githubUrl ?: "")
                                val repoKey = GithubRepoInfoService.getRepoKey(url)
                                if (repoKey != null) {
                                    val repoRes = GithubRepoInfoService.getRepoInfo(repoKey.owner, repoKey.repo)
                                    if (repoRes.result == Result.NONE) {
                                        dispatchedGithubFetch = true
                                        ProgressUtils.runBackground(
                                            file.project,
                                            "DepLens: Fetch GitHub ${repoKey.owner}/${repoKey.repo}"
                                        ) {
                                            GithubRepoInfoService.fetchRepoInfo(repoKey.owner, repoKey.repo) {
                                                if (vFile != null) {
                                                    UiUtils.refreshInlayHints(project, vFile)
                                                }
                                            }
                                        }
                                    }
                                } else {
                                    LOG.warn("Failed to get repo key for $pkg")
                                }
                            } else {
                                // 没有拿到元数据，插入错误
                            }

                            // npm 请求结束后至少刷新一次，避免状态停留在 loading。
                            if (!dispatchedGithubFetch) {
                                if (vFile != null) {
                                    UiUtils.refreshInlayHints(project, vFile)
                                }
                            }
                        }
                    } catch (e: Exception) {
                        LOG.warn("Failed to load repo info for $pkg", e)
                    }
                }
                UiUtils.addInlay(
                    sink,
                    offset,
                    I18n.message(I18nKey.loadingNpmMeta),
                    retryToken = "npm:$pkg"
                )
                return
            }
            Result.SUCCESS -> {}
            Result.PENDING -> {
                UiUtils.addInlay(
                    sink,
                    offset,
                    I18n.message(I18nKey.loadingNpmMeta),
                    retryToken = "npm:$pkg"
                )
                return
            }
            else -> {
                UiUtils.addInlay(
                    sink,
                    offset,
                    I18n.message(I18nKey.failedNpmMeta),
                    retryToken = "npm:$pkg"
                )
                return
            }
        }

        val npmInfo = npmRes.data ?: run {
            UiUtils.addInlay(
                sink,
                offset,
                I18n.message(I18nKey.failedNpmMeta),
                retryToken = "npm:$pkg"
            )
            return
        }
        val url = npmInfo.githubUrl
        if (url.isNullOrBlank()) {
            UiUtils.addInlay(
                sink,
                offset,
                I18n.message(I18nKey.noGithubUrl),
                retryToken = "npm:$pkg"
            )
            return
        }
        val repoKey = GithubRepoInfoService.getRepoKey(url)
        if (repoKey == null) {
            UiUtils.addInlay(
                sink,
                offset,
                I18n.message(I18nKey.invalidGithubUrl),
                retryToken = "npm:$pkg"
            )
            return
        }

        val repoRes = GithubRepoInfoService.getRepoInfo(repoKey.owner, repoKey.repo)

        val displayText = when (repoRes.result) {
            Result.NONE -> {
                ProgressUtils.runBackground(
                    file.project,
                    "DepLens: Fetch GitHub ${repoKey.owner}/${repoKey.repo}"
                ) {
                    GithubRepoInfoService.fetchRepoInfo(repoKey.owner, repoKey.repo) {
                        if (vFile != null) {
                            UiUtils.refreshInlayHints(project, vFile)
                        }
                    }
                }
                I18n.message(I18nKey.loadingGithub)
            }
            Result.PENDING -> I18n.message(I18nKey.loadingGithub)
            Result.SUCCESS -> "⭐ ${repoRes.data?.stars ?: 0} • ${I18n.message(I18nKey.lastUpdated)} ${repoRes.data?.updatedDate ?: "N/A"}"
            else -> I18n.message(I18nKey.failedGithub)
        }

        val githubUrl = "https://github.com/${repoKey.owner}/${repoKey.repo}"
        UiUtils.addInlay(
            sink,
            offset,
            displayText,
            githubUrl = githubUrl,
            retryToken = "github:${repoKey.owner}/${repoKey.repo}"
        )
    }
}

File: vscode/src/utils/inlay/NpmInlayUtils.ts
import type * as vscode from "vscode";
import { I18nKey } from "../../common/I18nKey";
import { Result } from "../../common/Result";
import { I18n } from "../I18n";
import { ProgressUtils } from "../ProgressUtils";
import { UiUtils } from "../UiUtils";
import { GithubRepoInfoService } from "../service/GithubRepoInfoService";
import { NpmPkgInfoService } from "../service/NpmPkgInfoService";

export class NpmInlayUtils {
  static addNpmDepInlay(hints: vscode.InlayHint[], pkg: string, position: vscode.Position): void {
    const npmRes = NpmPkgInfoService.getInstance().getPackageInfo(pkg);

    switch (npmRes.result) {
      case Result.NONE:
        if (
          NpmPkgInfoService.getInstance().hasFailure(pkg) &&
          !NpmPkgInfoService.getInstance().isRequestRunning(pkg)
        ) {
          UiUtils.addInlay(
            hints,
            position,
            I18n.message(I18nKey.failedNpmMeta),
            undefined,
            undefined,
            `npm:${pkg}`,
          );
          return;
        }

        void Promise.resolve(
          ProgressUtils.runBackground(`DepLens: Fetch npm ${pkg}`, () =>
            NpmPkgInfoService.getInstance().fetchPackageInfo(pkg),
          ),
        ).catch((error: unknown) => {
          console.warn(`Failed to load npm info for ${pkg}`, error);
        });
        UiUtils.addInlay(
          hints,
          position,
          I18n.message(I18nKey.loadingNpmMeta),
          undefined,
          undefined,
          `npm:${pkg}`,
        );
        return;

      case Result.PENDING:
        UiUtils.addInlay(
          hints,
          position,
          I18n.message(I18nKey.loadingNpmMeta),
          undefined,
          undefined,
          `npm:${pkg}`,
        );
        return;

      case Result.SUCCESS:
        break;

      case Result.FAILURE:
        UiUtils.addInlay(
          hints,
          position,
          I18n.message(I18nKey.failedNpmMeta),
          undefined,
          undefined,
          `npm:${pkg}`,
        );
        return;
    }

    const npmInfo = npmRes.data;
    if (!npmInfo) {
      UiUtils.addInlay(
        hints,
        position,
        I18n.message(I18nKey.failedNpmMeta),
        undefined,
        undefined,
        `npm:${pkg}`,
      );
      return;
    }

    const url = npmInfo.githubUrl;
    if (!url) {
      UiUtils.addInlay(
        hints,
        position,
        I18n.message(I18nKey.noGithubUrl),
        undefined,
        undefined,
        `npm:${pkg}`,
      );
      return;
    }

    const repoKey = GithubRepoInfoService.getRepoKey(url);
    if (!repoKey) {
      UiUtils.addInlay(
        hints,
        position,
        I18n.message(I18nKey.invalidGithubUrl),
        undefined,
        undefined,
        `npm:${pkg}`,
      );
      return;
    }

    const repoRes = GithubRepoInfoService.getInstance().getRepoInfo(repoKey.owner, repoKey.repo);
    switch (repoRes.result) {
      case Result.NONE:
        void Promise.resolve(
          ProgressUtils.runBackground(
            `DepLens: Fetch GitHub ${repoKey.owner}/${repoKey.repo}`,
            () => GithubRepoInfoService.getInstance().fetchRepoInfo(repoKey.owner, repoKey.repo),
          ),
        ).catch((error: unknown) => {
          console.warn(`Failed to load repo info for ${repoKey.owner}/${repoKey.repo}`, error);
        });
        UiUtils.addInlay(
          hints,
          position,
          I18n.message(I18nKey.loadingGithub),
          undefined,
          `https://github.com/${repoKey.owner}/${repoKey.repo}`,
          `github:${repoKey.owner}/${repoKey.repo}`,
        );
        return;

      case Result.PENDING:
        UiUtils.addInlay(
          hints,
          position,
          I18n.message(I18nKey.loadingGithub),
          undefined,
          `https://github.com/${repoKey.owner}/${repoKey.repo}`,
          `github:${repoKey.owner}/${repoKey.repo}`,
        );
        return;

      case Result.SUCCESS: {
        const stars = repoRes.data?.stars ?? "0";
        const updated = repoRes.data?.updatedDate ?? "N/A";
        UiUtils.addInlay(
          hints,
          position,
          `⭐ ${stars} • ${I18n.message(I18nKey.lastUpdated)} ${updated}`,
          undefined,
          `https://github.com/${repoKey.owner}/${repoKey.repo}`,
          `github:${repoKey.owner}/${repoKey.repo}`,
        );
        return;
      }

      case Result.FAILURE:
        UiUtils.addInlay(
          hints,
          position,
          I18n.message(I18nKey.failedGithub),
          undefined,
          `https://github.com/${repoKey.owner}/${repoKey.repo}`,
          `github:${repoKey.owner}/${repoKey.repo}`,
        );
        return;
    }
  }
}
